tf2_gnn/layers/graph_densely_connected_transformer.py

```diff
@@ -0,0 +1,219 @@
+import tensorflow as tf
+
+from tf2_gnn.utils.param_helpers import get_activation_function
+from dpu_utils.tf2utils import unsorted_segment_softmax
+
+
+# This helper function is currently not in use, as it is extremely slow
+# in the computation graph. Instead, we pre-compute the list of index pairs.
+def _all_shifted_pairs(lengths: tf.Tensor) -> tf.Tensor:
+    def _lengths_to_indices(lengths: tf.Tensor) -> tf.RaggedTensor:
+        return tf.ragged.range(
+            starts=tf.cumsum(
+                tf.concat([[0], lengths[:-1]], axis=0)
+            ),  # Starts of index ranges
+            limits=tf.cumsum(lengths),  # End of index ranges
+        )
+
+    def _cartesian_prod(a: tf.Tensor, b: tf.Tensor) -> tf.Tensor:
+        return tf.reshape(tf.stack(tf.meshgrid(a, b, indexing="ij"), axis=-1), (-1, 2))
+
+    indices = _lengths_to_indices(lengths)
+    return tf.map_fn(lambda x: _cartesian_prod(x, x), indices).flat_values
+
+
+class GraphDenselyConnectedTransformerLayer(tf.keras.layers.Layer):
+    """Transformer-like layer for sparsely batched graphs.
+    Applies a standard Transformer layer such that all nodes of each graph attend to
+    all (other and same) nodes in the same graph. This entirely ignores the graph
+    topology, and instead treats the node representations as a bag of inputs.
+    """
+    def __init__(
+        self,
+        num_heads: int = 6,
+        head_size: int = 16,
+        hidden_size: int = 64,
+        intermediate_size: int = 128,
+        intermediate_activation_name: str = "gelu",
+        hidden_dropout_rate: float = 0.1,
+        attention_probs_dropout_rate: float = 0.1,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        self._num_heads = num_heads
+        self._head_size = head_size
+        self._hidden_size = hidden_size
+        self._intermediate_size = intermediate_size
+        self._intermediate_act_fn = get_activation_function(
+            intermediate_activation_name
+        )
+        self._hidden_dropout_rate = hidden_dropout_rate
+        self._attention_probs_dropout_rate = attention_probs_dropout_rate
+
+        # Sublayer definitions:
+        self._node_label_to_qkv_layer = tf.keras.layers.Dense(
+            units=3 * (num_heads * head_size), activation=None, use_bias=False,
+        )
+        self._attention_to_pre_intermediate_layer = tf.keras.layers.Dense(
+            units=hidden_size, activation=None, use_bias=False,
+        )
+        self._pre_intermediate_layer_norm = tf.keras.layers.LayerNormalization(epsilon=1e-12)
```

Comment:
Layer Norm eps is quite small, so it caught my eye. I'm sure you have a good reason for making this so much smaller than the _Attention is all you need_ case, but then it may be good to add a note here.

Reply:
The `1e-12` is the TF1 default; the Keras/TF2 default is `1e-3`, which is high enough that it starts making a difference. I'm not particularly married to any constant here; I just picked `1e-12` because I knew it to not be an issue. (It's only used to avoid a division by 0 at one point, so it's influence is fairly limited beyond small numerical issues)

Reply:
Sounds good, thanks.